### DISCUSSION TITLE
Postfix expressions for Units of Measure using PrefixBind and PostfixBind methods

### DISCUSSION THREAD

**Discussion by marinasundstrom** [dotnet/csharplang#7031] [Language Ideas]
This proposals is about implementing syntactic support for postfix expressions for units (of measure) - like SI and Imperial units of measurements.

The purpose is to give the infrastructure and syntax that allows for third-party libraries for scientific calculations to tap into.

Based on Manifold's [proposal for adding unit expressions](https://github.com/manifold-systems/manifold/tree/master/manifold-deps-parent/manifold-ext#unit-expressions) to Java.

Check [my experiment](https://github.com/marinasundstrom/CsharpUnitExpressionsProposal/blob/main/UnitExpressionsProposal.Tests/UnitTest1.cs) (without syntax, obviously)

By structurally adding a ```PostfixBind``` method to the type of variable that acts as a prefix you could implement syntax for dealing with units.

The compiler would detect that there is a ```PostfixBind``` (or ```PrefixBind```) defined on the the unit.

Syntax proposal:

```csharp
using static MassUnit;

Mass m = 20 kg;
```

The lowered code:

```csharp
Mass m = MassUnit.kg.PostfixBind(20);
```

Whereas ```kg``` is a static readonly field in ```MassUnit```;

In essence, the unit acts as a decorator, in the form of a method taking the preceding value as an argument, and then wraps or transforms it into another form.

This is (roughly) how it would be implemented:

```csharp
public sealed class MassUnit
{
    public static readonly MassUnit kg = new MassUnit("kilograms", "kg");

    protected MassUnit(string name, string symbol) 
    {
        // Implementation omitted
    }

    public Mass PostfixBind(double magnitude)
    {
        // Create the Mass with the current unit that this type represents.

        return new Mass(magnitude, this);
    }
}

public struct Mass 
    : IComparable<Mass>, IConvertible, IEquatable<Mass>, IParsable<Mass>, System.Numerics.IAdditionOperators<Mass,Mass,Mass>
{
    public Mass(double magnitude, MassUnit unit) 
    {
        // Implementation omitted
    }

    public Mass ToPounds()  
    {
        if(Unit == MassUnit.lb) return Magnitude;

        var pounds = Magnitude * 2.20462262185

        return new Mass(pounds, MassUnit.lb);
    }
}
```

The ```Mass``` type would deal with conversion between values of different ```MassUnit```s. And other Units would do the same. 

In addition to ```PostfixBind``` methods, there would be support for ```PrefixBind``` which might be self-explanatory.

## More on expressions

Since units are part of expressions you can do arithmetic operations on units - thanks to the good old operator overloading.

```csharp
using static LengthUnit;
using static TimeUnit;

Speed s = 15 m/s;
```

The lowered code:

```csharp
Speed s = (LengthUnit.m / TimeUnit.s).PostfixBind(15);
```

**Comment by iam3yal** [comment]
It was pretty neat if we could put `PostfixBindAttribute` or `PrefixBindAttribute` on static methods and the compiler would support additional form of the invocation expression so instead of writing this `Mass.kg(20d)` in the traditional form you could rewrite it as `20d kg`, you can get close to it today through extension methods `20d.kg()` but it's kinda horrific.

**Reply by marinasundstrom** [reply]
Btw. I was thinking that the ```PrefixBind``` and ```PostfixBind``` could also be declared as extensions methods. Although that would not add so much value.

**Reply by iam3yal** [reply]
@marinasundstrom You're right it doesn't add much value when `PrefixBind` and `PostfixBind` are declared as extension methods, my thought though was declaring `kg` either as a static method or extension method and mark them with `PostfixBindAttribute` and `PrefixBindAttribute` respectively and extend the language to support this form of invocation.

So something like this:
```csharp
public struct Mass 
{
    [PostfixBind]
    public static Mass kg(double magnitude)
        => new Mass(magnitude);
}
```
Or
```csharp
public static class MassExtensions
{
    [PostfixBind]
    public static Mass kg(this double magnitude)
        => new Mass(magnitude);

    [PostfixBind]
    public static Mass kg(this int magnitude)
        => new Mass(magnitude);
}
```

**Comment by msedi** [comment]
Awesome idea. We currently need to do this via runtime parsing with all the negative side-effects (runtime errors, performance, parsing, etc)

`private PhysNumber myConcentration = "15e-3 g/cm^3"; `

Ultimately I would prefer a code rewriter (similar to the source generator), but I think that currently beyond all efforts.
On the other hand it would be easier if C# or .NET support such a type natively.

**Comment by marinasundstrom** [comment]
Check [my experiment](https://github.com/marinasundstrom/CsharpUnitExpressionsProposal/blob/main/UnitExpressionsProposal.Tests/UnitTest1.cs) (without syntax, obviously!)

**Comment by Richiban** [comment]
I love this idea! I got used to them in F# and really miss them now in C#.

Just a note on terminology; I'd recommend calling this "Units of measure" (to match F#) instead of "Unit expressions", because the latter is highly suggestive of the use of the `Unit` type in expressions, which is something else entirely.

**Reply by marinasundstrom** [reply]
Great point. Changed the title to "Postfix expressions for Units of Measure using PrefixBind and PostfixBind methods"

**Comment by Richiban** [comment]
Does your proposal include any type safety for differing units in the same dimension?

I.e. ideally the type checker would reject trying to assign a value of `miles` to a value of type `kilometres`.

**Reply by marinasundstrom** [reply]
I have not written about that, but depending on how you implement the units, which are classes, then the compiler would complain about incompatible types.

Ideally, there would be subtypes of, for instance, ```LengthUnit``` and conversion between the types.

**Reply by marinasundstrom** [reply]
Though it seems unlikely that you would be able to do something like this:

```chsharp
Speed value_mph = 20 miles/h;
var value_kmph = (km/h)value_mph;
````

Since "km/h" would be "constants" and not types.

**Reply by Richiban** [reply]
F# has the wonderful feature of actually multiply / divide the units together to produce new units on the fly. For example:

```fs
[<Measure>] type mi
[<Measure>] type km
[<Measure>] type h 

let speedInMph = 30.0<mi/h>

let conversionRate = 1.60934<km/mi>

let speedInKph = speedInMph * conversionRate // Automatically inferred as km/h
```

It's impossible to pass a value of `km/h` to a function that expects `mi/h`, for example.

**Reply by marinasundstrom** [reply]
That could be enforced by analyzers in C#.

**Reply by msedi** [reply]
@Richiban. The question is if this is really a problem? What we did was that we translated things internally to the basic SI-units, so km/h and mi/h, is in the end length/time. 

You could also argue that m/h or m/s should also not match. I assume this is a matter of taste if this should be allowed or not. If you remember physics courses, you should first translate to a unique unit system so that you won't get into trouble. So this would prefer your way of doing it.

I'm unsure about it.

**Reply by HaloFour** [reply]
> I have not written about that, but depending on how you implement the units, which are classes, then the compiler would complain about incompatible types.

While this doesn't need to be considered in any early explorations, I don't think a language feature would be considered here unless it was a very low cost abstraction.  Requiring allocations between each conversion and operation would probably be considered too expensive.  That would rule out class hierarchies for defining the unit types.  Single-value structs would be a better strategy as the runtime understands how to elide them if they contain a single field.

**Reply by msedi** [reply]
> That would rule out class hierarchies for defining the unit types


I agree. A remark though, since thats something we see quite often with our own PhysNumbers is that you have a large structure of physical values where you canot directly see what unit the property has, e.g.

```cs
  public readonly struct MetricAcquisitionParameters : IReadOnlyMetricAcquisitionParameters
    {
        public PhysNumber DataBeginPos { get; }
        public PhysNumber DataEndPos { get; }
        public PhysNumber TableFeedPerRotation { get; }
        public PhysNumber Penumbra { get; }
        public PhysNumber FeedScan { get; }
        public PhysNumber GantryTilt { get; }
        public PhysNumber RotationTime { get; }
}
```

People don't know what unit for example `Penumbra` or `GantryTilt` have. Here it would be better to have a variety of physical value containers, like `Angle (GantryTilt)` or `Length (Penumbra)`. Having a simple `Unit` type would not support it and you could assign to `Angle`, e.g. 20 m/s. So you would need analyzers and attributes to tell the system what unit you are referring to. What you would wish here 

1. is derives from the Unit type which would not be possible,
_or_
2. there is a possibility that takes the Unit as a template for a source generator(?) that creates these pseudo derivatives.

**Comment by DavidArno** [comment]
With the plans around extensions being looked at by the team, one non-perfect solution to this would be to implement it all as extension properties:
```
Speed s = 15.m/1.s;
Concentration myConcentration = 15e-3.g/1.cm^3; 
```
It's not perfect syntax, but has the big advantage of not needing any changes to the C# language beyond an already in-progress change.

**Reply by marinasundstrom** [reply]
That is one solution.

The chances are low that they will introduce the prefix/postfix expression syntax without a clear need either within the framework of in the community.

**Reply by marinasundstrom** [reply]
Btw. I was thinking that the PrefixBind and PostfixBind could also be declared as extensions methods. Although that would not add so much value.

**Comment by marinasundstrom** [comment]
Wonder how hard it is to implement in Roslyn. 

But I guess that once either a literal or identifier expression has been parsed we should just check for an expression coming after. But when it comes to postfix, I'm unsure. We really don't know the meanings of the expressions until semantic analysis:

https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs#L10659

![Screenshot 2023-03-06 at 13 10 32](https://user-images.githubusercontent.com/919580/223106880-46639d99-f5f0-45b9-bd3f-7f752ea85ed0.png)

**Reply by msedi** [reply]
That's an interesting question, if roslyn code generation can be extended by a similar mechanism like source generators?

**Reply by marinasundstrom** [reply]
I was thinking about forking and building a prototype, since I know some about compilers. Making the parsing work must be pretty simple. The rest I'm unsure of-

It can't be extended with extensions like with Analyzers and Code fixes. That is by design. The designers don't know how to do it. And it could potentially cause confusion if someone implements their own syntax extension, it has bugs, and the compiler team gets the blame.

**Reply by msedi** [reply]
@marinasundstrom. I somehow understand that and fully agree, nevertheless it would be nice if it would be possible ;-)

**Reply by marinasundstrom** [reply]
On the topic, Nemerle is a .NET language which has macros that can be used to extend the language with syntax and behavior.

http://nemerle.org/About

**Comment by saint4eva** [comment]
A library like this could help:

[https://github.com/angularsen/UnitsNet](https://github.com/angularsen/UnitsNet)

**Reply by marinasundstrom** [reply]
Yes. I was thinking that it could work with the syntax that I proposed. It would just be a matter of adding the methods to support the syntax. The rest is up to the library.

```csharp
var kg = LengthUnit.Kg;

Length = 2 kg;
```

It is implemented as an extensions method:

```csharp
public static class LengthUnitExtensions
{
    public static Length PostfixBind(this LengthUnit unit, double magnitude)
    {
        return new Length(magnitude, unit);
    }
}
```

But in this model that the Units.NET library provides (with enums) the postfix expression part would be limited. Since you cannot express units as an expression in the form ```2 m/s``` but rather 

```csharp
var meterPerSecond = SpeedUnit.MeterPerSecond;

Speed = 15 meterPerSecond;
```

There are probably ways that Units.NET could support a more expressive postfix expression part if they would want to. Or consumers could write their own stuff for that.
